flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchSinkBuilder.java

```diff
@@ -0,0 +1,276 @@
+     *     requests, in milliseconds.
+     * @return {@link ElasticsearchSinkBuilder}
+     */
+    public ElasticsearchSinkBuilder<IN> setBulkFlushBackoffDelay(long delayMillis) {
+        checkState(
+                delayMillis >= 0,
+                "Delay (in milliseconds) between each backoff attempt must be larger "
+                        + "than or equal to 0.");
+        this.bulkFlushBackOffDelay = delayMillis;
+        return this;
+    }
+
+    /**
+     * Sets the username used to authenticate the connection with the Elasticsearch cluster.
+     *
+     * @param username of the Elasticsearch cluster user
+     * @return {@link ElasticsearchSinkBuilder}
+     */
+    public ElasticsearchSinkBuilder<IN> setConnectionUsername(String username) {
+        checkNotNull(username);
+        this.username = username;
+        return this;
+    }
+
+    /**
+     * Sets the password used to authenticate the conection with the Elasticsearch cluster.
+     *
+     * @param password of the Elasticsearch cluster user
+     * @return {@link ElasticsearchSinkBuilder}
+     */
+    public ElasticsearchSinkBuilder<IN> setConnectionPassword(String password) {
+        checkNotNull(password);
+        this.password = password;
+        return this;
+    }
+
+    /**
+     * Sets a prefix which used for every REST communication to the Elasticsearch cluster.
+     *
+     * @param prefix for the communication
+     * @return {@link ElasticsearchSinkBuilder}
+     */
+    public ElasticsearchSinkBuilder<IN> setConnectionPathPrefix(String prefix) {
+        checkNotNull(prefix);
+        this.connectionPathPrefix = prefix;
+        return this;
+    }
+
+    /** @return {@link ElasticsearchSink} */
+    @SuppressWarnings("unchecked")
+    public ElasticsearchSink<IN> build() {
+        checkNotNull(hosts);
+        checkNotNull(processor);
+        return new ElasticsearchSink<IN>(
+                hosts,
+                (ElasticsearchProcessor<IN>) processor,
+                deliveryGuarantee,
+                buildBulkProcessorConfig(),
+                new NetworkClientConfig(username, password, connectionPathPrefix));
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T extends IN> ElasticsearchSinkBuilder<T> self() {
+        return (ElasticsearchSinkBuilder<T>) this;
+    }
+
+    private BulkProcessorConfig buildBulkProcessorConfig() {
+        if (bulkFlushBackoffType == null) {
+            checkState(
+                    bulkFlushBackoffRetries == -1,
+                    "Max number of backoff attempts cannot be used without backoff type.");
+            checkState(
+                    bulkFlushBackOffDelay == -1,
+                    "Delay (in milliseconds) between each backoff cannot be used "
+                            + "without backoff type.");
+            return new BulkProcessorConfig(bulkFlushMaxActions, bulkFlushMaxMb, bulkFlushInterval);
```

Comment:
Would be nice if we could enforce that with sub-builders. See Avro's SchemaBuilder for an example.